### PR TITLE
prov/sm2: Add CMA and CUDA IPC capabilities to sm2

### PR DESCRIPTION
**Pull request by sunkuamzn**
AWS needs SM2 to have the capability to send any message size in order to perform benchmarking on real world applications.

The CUDA IPC commit only enables device-device and device-host transfers using CUDA IPC.

~~Host-device transfers need a new protocol which will come later.~~ Added host to CUDA device commit

### CODE REVIEW

**Comment by wzamazon** [comment]
Need rebase to fix CI failure.

**Comment by sunkuamzn** [comment]
@tschuett I cannot find that flag https://github.com/search?q=repo%3Aofiwg%2Flibfabric%20FI_CUDA_API_PERMITTED&type=code

Can you elaborate?

**Comment by tschuett** [comment]
See this discussion:
https://github.com/ofiwg/libfabric/issues/8610
and PR: https://github.com/ofiwg/libfabric/pull/8624

File: prov/sm2/src/sm2_ep.c
Comment (a-szegel) [inline prov/sm2/src/sm2_ep.c:350]: I think you need to add delivery complete here.

File: prov/sm2/src/sm2_msg.c
Comment (a-szegel) [inline prov/sm2/src/sm2_msg.c:318]: Line 322... this is as close as it will let me make a comment on...  We need to add sm2_proto_ipc here... Another better alternative solution is to pass op_flags as a pointer, so we don't need to keep extending it... if you choose to do it this way, please make the initial change for the other protocols in a separate commit.
```
if (!(op_flags & FI_DELIVERY_COMPLETE) && proto != sm2_proto_cma) {
```

**Review by a-szegel — CHANGES_REQUESTED** [review]

File: prov/sm2/src/sm2_ep.c
Comment (wenduwan) [inline prov/sm2/src/sm2_ep.c:837]: Reminder, we fixed a bug on read path. We cannot use inject in that case.

https://github.com/ofiwg/libfabric/pull/8925

File: prov/sm2/src/sm2_attr.c
Comment (a-szegel) [inline prov/sm2/src/sm2_attr.c:100]: You will need to undo this change. in your PR... how did you test this?

**Comment by vidsouza** [comment]
bot:aws:retest

File: prov/sm2/src/sm2_domain.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_domain.c]: fixed

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c]: fixed

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c]: I added this according to SHM before your CMA change. I fixed it now.

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c]: fixed

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c]: fixed

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c]: fixed

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c:414]: fixed

File: prov/sm2/src/sm2_progress.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c]: i removed it

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c:364]: fixed

File: prov/sm2/src/sm2_msg.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_msg.c:354]: went with the pointer approach and put that in a separate commit

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c]: we don't have RMA yet

File: prov/sm2/src/sm2_attr.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_attr.c:100]: keeping it because we don't have large message device-host

File: prov/sm2/src/sm2_ep.c
Comment (a-szegel) [inline prov/sm2/src/sm2_ep.c]: So I know I asked you to do this... but can you actually re-base this change in front of the cma change, and then squash the cma specific part into the cma commit?  This will provide a cleaner development history for a reader.

Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c]: done

**Review by a-szegel — APPROVED** [review]
LGTM... tyvm :)

File: prov/sm2/src/sm2_ep.c
Comment (lrbison) [inline prov/sm2/src/sm2_ep.c:341]: Don't forget to use the new [sm2_pop_xfer_entry](https://github.com/ofiwg/libfabric/pull/8941) function

Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c]: done

**Comment by sunkuamzn** [comment]
Some comments on the device to host commit are on https://github.com/ofiwg/libfabric/pull/8956

**Comment by a-szegel** [comment]
Looks like this change causes a 1-2us second regression in osu_alltoallw (minor), but osu_latency is fine.

File: prov/sm2/src/sm2.h
Comment (a-szegel) [inline prov/sm2/src/sm2.h:195]: this is wild but the osu_alltoallw degredation seems to come from this patch.

**Review by a-szegel — CHANGES_REQUESTED** [review]
This feels very "hacked in".  We add ~4 new if statements to the critical path of the host memory case.  This is done b/c we are trying to re-use the `sm2_progress_recv` -> `sm2_progress_recv_msg` -> `sm2_start_common` abstraction that we inherited from SHM.  I don't really think this holds for us anymore, and I would like to start thinking about a new abstraction.

My major concerns are:
1. I don't want to impact the performance of the non-cuda path (which honestly doesn't look that bad, but it isn't non-zero)
2. I want the HMEM stuff to be better isolated behind protocols, then big if statements in common receive path utilities
3. I want the code to look a little cleaner.

**Comment by a-szegel** [comment]
Can you run this with validation for a lot of iterations in an until_fail script?

**Comment by sunkuamzn** [comment]
bot:aws:retest

**Comment by sunkuamzn** [comment]
bot:aws:retest

**Comment by sunkuamzn** [comment]
bot:aws:retest

**Comment by sunkuamzn** [comment]
bot:aws:retest

File: prov/sm2/src/sm2.h
Comment (a-szegel) [reply prov/sm2/src/sm2.h:195]: I would ignore this for now.

File: prov/sm2/src/sm2_progress.c
Comment (a-szegel) [reply prov/sm2/src/sm2_progress.c:473]: after following back... this is low priority

**Comment by sunkuamzn** [comment]
bot:aws:retest

File: prov/sm2/src/sm2.h
Comment (lrbison) [reply prov/sm2/src/sm2.h:195]: I don't understand why the `sm2_generic_format` function should modify op_flags, and it doesn't seem to here or in your final commit.

**Comment by sunkuamzn** [comment]
@aingerson are the Intel CI failures real?

**Comment by aingerson** [comment]
@sunkuamzn The CI failure was the tcp tagged bw failure that has been popping up recently. Sean is looking into it. By the way, our new CI will display which step actually failed so you can see that it was tcp (the other two are the parent steps of the test, I will see if we can just print the one).

**Review by a-szegel — CHANGES_REQUESTED** [review]
final comments, then I am good

File: prov/sm2/src/sm2_progress.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c]: there's actually not much here. I could move the assert statement inside `sm2_progress_ipc_host_to_dev` but I would have to pass in `cma_data` to that function..

**Review by a-szegel — APPROVED** [review]

File: prov/sm2/src/sm2.h
Comment (lrbison) [reply prov/sm2/src/sm2.h:179]: That is in `sm2_do_cma`.  In `sm2_generic_format` the op_flag of the caller should not be modified, but passing a pointer to the int suggests it can be.

I will propose a separate follow-up later, I don't think this small nit needs to hold up this merge.

File: prov/sm2/src/sm2_progress.c
Comment (lrbison) [inline prov/sm2/src/sm2_progress.c:176]: I think this should test `errno == EFAULT` rather than 14.

**Review by lrbison — APPROVED** [review]

File: prov/sm2/src/sm2_progress.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c:176]: Fixed

File: prov/sm2/src/sm2.h
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h]: Fixed

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:60]: Please add a similar ofi_is_addr_valid(iface...) call in ofi_hmem.h so we are consistent with how we call into the hmem ops array).

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:61]: Also unnecessary comment

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:176]: Seems dramatic to try CMA and fail. We know the iface for the destination at this point so we don't have to fail CMA to fallback. We can just check rx_entry->desc (ofI_mr *) and fallback before calling the cma_loop function

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:198]: drop else, always returns in if

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:45]: This name seems misleading. This sounds like you are processing a xfer_entry that had proto = ipc_host_to_dev which is not what you're doing here. This is where you send back the entry to fallback to the IPC call.

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:105]: drop else

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:112]: Can we reuse/consolidate this with the normal ipc path? Same with the get base addr/open path. Also some of these names are getting long. Consolidating would help simplify them/make them more generic.

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:169]: how does the fallback IPC protocol deal with multiple target buffers?

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:365]: drop brackets

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:420]: Personally, I would drop the comment entirely since that's the definition of this protocol. If an explanation needs to go anywhere, it should just be where the protocol is defined/in the header.

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:434]: drop comment

File: prov/sm2/src/sm2_progress.c
Comment (aingerson) [inline prov/sm2/src/sm2_progress.c:439]: Again, this is definition of the protocol itself, so I would drop the comment.

**Review by aingerson — COMMENTED** [review]
I would add more in each commit message (generally speaking, I would say no commit should be blank unless it's absolutely self-explanatory)
In general, my biggest suggestion would be to rework the op_flags (setting them and unsetting them in order to trigger/avoid completions in specific places) and also streamlining the overlap between the normal ipc path and fallback ipc, especially in terms of how we pass variables around or how we interpret the user_data. I think ideally we reduce the number of ways we can interpret the user_data so we're not casting it left and right. The whole xfer_entry is 4k and we don't need any extra for the payload for cma/ipc so I don't think there's much harm in including all the variables we might need for cma+ipc in case of the fallback so we don't have to move the cma iov ptr to the host pointer and other weird adjustments. I think it could be simplified by putting it all into one, sending the CMA, if the target is on the device, format the ipc section and send it back to the sender with almost exactly what they sent, just with the ipc info also. I think that would make it a lot easier to follow and read.

File: prov/sm2/src/sm2.h
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:214]: this was a mistake. fixed

File: prov/sm2/src/sm2_ep.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_ep.c:244]: That's a great idea. I can make a separate PR for that after this is merged.

File: prov/sm2/src/sm2.h
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:170]: i'm using just rx_entry, like you suggested in a different comment. rx_flags are needed when generating the receive completion.

File: prov/sm2/src/sm2_progress.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c:176]: fixed

File: prov/sm2/src/sm2_progress.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c:169]: thanks for catching. i've added this.

File: prov/sm2/src/sm2.h
Comment (sunkuamzn) [reply prov/sm2/src/sm2.h:101]: I've addressed this by getting rid of the two new protocols and instead using flags in the header. I split the 32 bit proto into 16 bit proto and 16 bit proto_flags. So now instead of setting a different protocol every time, I just change the flags.

This does make the recv path a bit messy but we are planning a refactor where we switch on protocol before switching on op and that would make this a lot cleaner.

File: prov/sm2/src/sm2_progress.c
Comment (sunkuamzn) [reply prov/sm2/src/sm2_progress.c:60]: not using this function anymore, as we discussed

**Comment by sunkuamzn** [comment]
bot:aws:retest

**Comment by sunkuamzn** [comment]
Continued in https://github.com/ofiwg/libfabric/pull/9103

**Pull request closed without merging**
